channels/rdpecam/client/v4l/camera_v4l.c

```diff
@@ -0,0 +1,731 @@
+			deviceId = (char*)cap.bus_info;
+		else
+			deviceId = device;
+
+		IFCALL(callback, ecam, hchannel, deviceId, deviceName);
+
+		close(fd);
+	}
+
+	return count;
+}
+
+/**
+ * Function description
+ *
+ * @return void
+ */
+static void cam_v4l_stream_free_buffers(CamV4lStream* stream)
+{
+	if (!stream || !stream->buffers)
+		return;
+
+	/* unmap buffers */
+	for (int i = 0; i < stream->nBuffers; i++)
+	{
+		if (stream->buffers[i].length && stream->buffers[i].start != MAP_FAILED)
+		{
+			munmap(stream->buffers[i].start, stream->buffers[i].length);
+		}
+	}
+
+	free(stream->buffers);
+	stream->buffers = NULL;
+	stream->nBuffers = 0;
+	return;
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on failure, otherwise allocated buffer size
+ */
+static size_t cam_v4l_stream_alloc_buffers(CamV4lStream* stream)
+{
+	struct v4l2_requestbuffers rbuffer;
+	memset(&rbuffer, 0, sizeof(rbuffer));
+
+	rbuffer.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+	rbuffer.memory = V4L2_MEMORY_MMAP;
+	rbuffer.count = CAM_V4L2_BUFFERS_COUNT;
+
+	if (ioctl(stream->fd, VIDIOC_REQBUFS, &rbuffer) < 0 || rbuffer.count == 0)
+	{
+		WLog_ERR(TAG, "Failure in VIDIOC_REQBUFS, errno %d, count %d", errno, rbuffer.count);
+		return 0;
+	}
+
+	stream->nBuffers = rbuffer.count;
+
+	/* Map the buffers */
+	stream->buffers = (CamV4lBuffer*)calloc(rbuffer.count, sizeof(CamV4lBuffer));
+	if(!stream->buffers)
+	{
+		WLog_ERR(TAG, "Failure in calloc");
+		return 0;
+	}
+
+	for (unsigned int i = 0; i < rbuffer.count; i++)
+	{
+		struct v4l2_buffer buffer;
+		memset(&buffer, 0, sizeof(buffer));
+		buffer.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+		buffer.memory = V4L2_MEMORY_MMAP;
+		buffer.index = i;
+
+		if (ioctl(stream->fd, VIDIOC_QUERYBUF, &buffer) < 0)
+		{
+			WLog_ERR(TAG, "Failure in VIDIOC_QUERYBUF, errno %d", errno);
+			cam_v4l_stream_free_buffers(stream);
+			return 0;
+		}
+
+		stream->buffers[i].start = mmap(NULL, buffer.length, PROT_READ | PROT_WRITE,
+						MAP_SHARED, stream->fd, buffer.m.offset);
+
+		if (MAP_FAILED == stream->buffers[i].start)
+		{
+			WLog_ERR(TAG, "Failure in mmap, errno %d", errno);
+			cam_v4l_stream_free_buffers(stream);
+			return 0;
+		}
+
+		stream->buffers[i].length = buffer.length;
+
+		WLog_DBG(TAG, "Buffer %d mapped, size: %d", i, buffer.length);
+
+		if (ioctl(stream->fd, VIDIOC_QBUF, &buffer) < 0)
+		{
+			WLog_ERR(TAG, "Failure in VIDIOC_QBUF, errno %d", errno);
+			cam_v4l_stream_free_buffers(stream);
+			return 0;
+		}
+	}
+
+	return stream->buffers[0].length;
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT cam_v4l_stream_capture_thread(void* param)
+{
+	CamV4lStream* stream = (CamV4lStream*)param;
+
+	int fd = stream->fd;
+
+	do
+	{
+		int retVal;
+		fd_set rSet;
+		struct timeval timeout;
+
+		FD_ZERO(&rSet);
+		FD_SET(fd, &rSet);
+		timeout.tv_sec = 1;
+		timeout.tv_usec = 0;
+
+		retVal = select(fd + 1, &rSet, NULL, NULL, &timeout);
```

Comment:
better use `poll` that supports fd > 1024

Reply:
Done